### PR TITLE
Add Basic Auth Support

### LINKED ISSUE #902: Authentication for HTTP server
There could be an endpoint to login in order to post streams. The credentials could be passed as env variables.
`url:4195/login`

### PR DESCRIPTION
**Pull request by kennyp**
Adds HTTP BasicAuth support to the API server.

Closes #902 

While adding docs for `basic_auth` I pulled in the `Common` vs. `Advanced` tab layout.

### CODE REVIEW

**Review by Jeffail — COMMENTED** [review]
Thanks @kennyp, looks good, added a few comments for stuff that'd make it a bit more solid.

File: internal/http/docs/basic_auth.go [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+// BasicAuthFieldSpec returns the spec for an HTTP BasicAuth component
+func BasicAuthFieldSpec() docs.FieldSpec {
+	return docs.FieldObject("basic_auth", "Require HTTP Basic Auth for the HTTP server.").WithChildren(
+		docs.FieldString("username", "Basic Auth Username").HasDefault(""),
+		docs.FieldString("password", "Basic Auth Password").HasDefault(""),
+	).Advanced()
+}
+
+func (b BasicAuth) enabled() bool {
+	return b.Username != "" && b.Password != ""
```
[the remark this answers is not in the document]
Comment (kennyp) [reply internal/http/docs/basic_auth.go:54]: Good call!

File: internal/http/docs/basic_auth.go [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+}
+
+// WrapHandler wraps the provided HTTP handler with middleware that enforces
+// BasicAuth if it's enabled.
+func (b BasicAuth) WrapHandler(next http.HandlerFunc) http.HandlerFunc {
+	if !b.Enabled {
+		return next
+	}
+
+	return func(next http.HandlerFunc, auth BasicAuth) http.HandlerFunc {
```
Comment (ekeric13) [inline internal/http/docs/basic_auth.go:81]: What's the value of this additional closure? It doesn't seem like anything passed is dynamic and therefore needs to be passed by value.

Comment (kennyp) [reply internal/http/docs/basic_auth.go:81]: In and earlier iteration I was closing over some other stuff. I think it should be fine to remove it. Thanks for looking out!

**Comment by Jeffail** [comment]
Looks good, thanks @kennyp! (and @ekeric13 for the review :heart:)

**Pull request merged**